### ISSUE TITLE
Addition of Metric Learning Heads like ArcFace , CosFace ,etc

### ISSUE THREAD

**Issue by Atharva-Phatak** [PyTorchLightning/lightning-bolts#603]
## 🚀 Feature

Metric Learning (especially person re-id) is an important research area. What I want is to add metric learning modules like ArcFace, CosFace in bolts so that it will be easy for researchers or people who want to use metric learning-based modules in their pipeline

### Motivation

Personally, whenever I work with metric learning modules I always have to code them up and it's frustrating to code everything from scratch(now I am quite comfortable). I do not want to implement them again and again and I think PL-Bolts is a nice place to have SOTA metric learning modules implemented.

### Pitch

To put it simply I want to implement metric learning modules so that they can be used with pretrained vision models for varied metric learning tasks(re-id , image-recognition, etc)

### Additional Context

I would like to personally implement and integrate these modules in PL-Bolts.

**Comment by akihironitta** [comment]
@SkafteNicki might be interested in this?

**Comment by SkafteNicki** [comment]
@akihironitta that depends on the scope here. @Atharva-Phatak are we talking about the full module or only the loss functions?

**Comment by Atharva-Phatak** [comment]
@SkafteNicki I think loss functions should be better for now, since the backbone construction is most of the time custom depending upon the problem. But usage of arc face loss and cos face loss seems to be the same. Of course in future versions, we can add full modules with i-resnet and stuff. But I would like to know your thoughts as well

**Comment by SkafteNicki** [comment]
I am asking, because maybe if we are only talking about the loss functions it may fit better within the scope of our [torchmetrics](https://github.com/PyTorchLightning/metrics) package. That said, bolts also have losses implemented so I am also fine with the implementations staying here.

**Comment by Atharva-Phatak** [comment]
I thought bolts would be a better place since most research stuff is implemented in bolts, so that's why. But I am fine with it implemented anywhere. Should I create a pull request ? If so where should I add them in pl-bolts ?

Implementation is pretty simple as well. Very Simple classes with some core logic. 

@SkafteNicki @akihironitta would like some guidance.

**Comment by SkafteNicki** [comment]
Then lets go for bolts :]
Is there any specific reason that they need to be implemented as classes or could they be simple functions?
IMO i would say:
* create a single file called `metric_learning.py` in folder `pl_bolts/losses/`
* implement all loss functions in that file
* implement corresponding test in `tests/losses/test_metric_learning.py` where you make sure the metrics calculate exactly what they are supposed to do. Maybe make some handcrafted examples?

I assume that you can begin implementing, but @akihironitta option is also important since he is the main responsible for bolts and I am for torchmetrics.

**Comment by Atharva-Phatak** [comment]
Well, the main reason for implementing them as classes is because you need to initialize the weight vector-only with some random weights, as our model trains the respective weights get updated with the backdrop, its not in its true sense a loss function but more like some geometric operations done on the input features(output of the backbone) to increase inter-class distance and minimize the intraclass distance on a manifold space. 

The ArcFaceHead usually returns raw logits, but we can include an argument for criterion in the class so that instead of returning logits we can return loss.

Here is a dummy example how it is usually used

```
class Model:
    def __init__(self , some_params_for_model , some_params_for_Arcface):
         self.model = some_model
         self.metric_head = ArcFace(some_params) #metric learning head
         self.classification_head = nn.Linear(some_shape , some_output_shape)

    def forward(self , x , labels): 
          # The forward pass requires the labels for computing the margins.
          model_op = self.model(x)
          metric_op = self.metric_head(model_op , labels) # this will return logits
          classification_op = self.classification_head(model_op)
          
          return classification_op , metric_op

model = Model(some_params)
classifcation_op , metric_op = model(x , labels)

metric_loss = nn.CrossEntropyLoss()(metric_op , labels)
cls_loss = nn.CrossEntropyLoss()(classifcation_op , labels)

loss = metric_loss + cls_loss
```

The above is a standard way how it is used in research. Its not a loss per see but a custom head to standard vision models. Since its job is to optimize the embedding space with the help of a standard loss function(could be any).
